### PR TITLE
Implemented node loading from JSON configuration file.

### PR DESCRIPTION
**Pull request by ddobric**
## Purpose
Changes proposed in this PR include additional feature, which enables loading of nodes from JSON configuration file. All hard-coded nodes remain included in simulator. Nodes loaded from JSON are NOT included in simulation. This feature populates nodes from JSON, so other applications (OPCUA Clients) can write and read values.
## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[X] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

**Comment by ddobric** [comment]
@hansgschossmann just as reminder. :)

**Comment by hansgschossmann** [comment]
@ddobric Sorry for not looking earlier to it. I will complete it by next week.

File: README.md [hunk 1]
```diff
@@ -12,6 +12,10 @@ The following nodes are part of the PLC simulation:
 - a value showing a negative trend
 
 By default everything is enabled, please use command line options to disable certain anomaly or data generation features.
+Additionally to simulation mode, a JSON configuration file, which defines nodes to be populated.
```
Comment (hansgschossmann) [inline README.md]: >file [](start = 54, length = 4)

file can be specified,

File: README.md [hunk 2]
```diff
@@ -47,6 +51,52 @@ X.509 certificates releated:
 * Running as Linux docker container, you can map the certificate stores to the host file system by using the docker run option `-v <hostdirectory>:/appdata`. This will make the certificate persistent over starts.
 * Running as Linux docker container and want to use an X509Store for the application certificate, you need to use the docker run option `-v x509certstores:/root/.dotnet/corefx/cryptography/x509stores` and the application option `--at X509Store`
 
+## Node population from JSON configuration file
+If the module (application) is started with argument **--nodesfile** then specified JSON configuration file is loaded.
```
Comment (hansgschossmann) [inline README.md]: >then specified JSON [](start = 69, length = 19)

then specified JSON -> then the specified JSON

File: README.md [hunk 3]
```diff
@@ -47,6 +51,52 @@ X.509 certificates releated:
...
+Nodes defined in JSON file will not be exposed by server. This enables another OPCUA-client application to set the state/value of the node. Please note that nodes specified in JSON file are NOT a part of simulation. They remain visible in unchanged state as long some OPCUA-client application changes their status.
```
Comment (hansgschossmann) [inline README.md]: >exposed by server [](start = 39, length = 17)

what do you mean my "exposed by server"? this is a bit unclear to me, can you change it?

File: README.md [hunk 4]
```diff
@@ -47,6 +51,52 @@ X.509 certificates releated:
...
+     
+      }
+    ]
+  }
+}
+~~~
+
+*NodeId*: Specifies the identifier of the node.
+*Name*: The name of the sensor.
+*DataType*: The OPC UA valid type. It specifies one of types defined by BuiltInType.
```
Comment (hansgschossmann) [inline README.md]: Is this the DisplayName or BrowseName?

File: README.md [hunk 5]
```diff
@@ -47,6 +51,52 @@ X.509 certificates releated:
...
+        "Description": "My property 2"
+     
+      }
+    ]
+  }
+}
+~~~
+
+*NodeId*: Specifies the identifier of the node.
+*Name*: The name of the sensor.
```
Comment (hansgschossmann) [inline README.md]: You should add that this is the NodeId in the NamespaceId

File: README.md [hunk 6]
```diff
@@ -47,6 +51,52 @@ X.509 certificates releated:
...
+~~~
+dotnet opcplc.dll --at X509Store --nodesfile nodesfile.json
+~~~
+Configuration file looks as described below:
+~~~
+{
+  "$type": "OpcPlc.ConfigFolder, opcplc",
+  "NamespaceId": 7,
+  "Name": "MyTelemetry",
+  "NodeList": {
```
Comment (hansgschossmann) [inline README.md]: you have not explained what those are for.

File: opcplc/OpcApplicationConfiguration.cs [hunk 7]
```diff
@@ -33,7 +33,7 @@ public static string Hostname
         public static string ApplicationName => ProgramName;
         public static string ApplicationUri => $"urn:{ProgramName}:{HostnameLabel}{(string.IsNullOrEmpty(ServerPath) ? string.Empty : (ServerPath.StartsWith("/") ? string.Empty : ":"))}{ServerPath.Replace("/", ":")}";
         public static string ProductUri => $"https://github.com/azure-samples/iot-edge-opc-plc";
-        public static ushort ServerPort { get; set; } = 50000;
+        public static ushort ServerPort { get; set; } = 50001;
         public static string ServerPath { get; set; } = string.Empty;
```
Comment (hansgschossmann) [inline opcplc/OpcApplicationConfiguration.cs]: what was the reason for this change? there are other samples, which rely on the value 50000 as default.

File: opcplc/PlcNodeManagerFromFile.cs [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+        private void CreateBaseVariable(FolderState root, ConfigNode node)
+        {
+            BuiltInType tp;
+            if (Enum.TryParse<BuiltInType>(node.DataType, out tp) == false)
+                throw new ArgumentException($"Value {node.DataType} of node {node.NodeId} cannot be parsed.");
+
+            byte accessLevel = 0;
+
+            // We have to hard code conversion here, because AccessLevel is defined as byte in OPCUA lib.
+            switch (node.AccessLevel)
```
Comment (hansgschossmann) [inline opcplc/PlcNodeManagerFromFile.cs]: How about using something like this:
            try
            {
                accessLevel = (byte)(typeof(AccessLevels).GetField(node.AccessLevel).GetValue(null));
            }
            catch (Exception e)
            {
                throw new ArgumentException($"Unsupported AccessLevel '{node.AccessLevel}'");
            }

File: opcplc/nodesfile.json [hunk 9]
```diff
@@ -0,0 +1,64 @@
+{
+  "$type": "OpcPlc.ConfigFolder, opcplc",
+  "NamespaceId": 7,
```
Comment (hansgschossmann) [inline opcplc/nodesfile.json]: pls remove

File: opcplc/nodesfile.json [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  "Name": "MyTelemetry",
```
Comment (hansgschossmann) [inline opcplc/nodesfile.json]: this is not picked up and created  at least when i was running. either you debug and fix it or remove it.

File: opcplc/nodesfile.json [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  "NodeList": {
+    "$type": "System.Collections.Generic.List`1[[OpcPlc.ConfigNode, opcplc]], System.Private.CoreLib",
+    "$values": [
```
Comment (hansgschossmann) [inline opcplc/nodesfile.json]: pls remove

File: opcplc/nodesfile.json [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      {
```
Comment (hansgschossmann) [inline opcplc/nodesfile.json]: just make NodeList the array and it is working

File: opcplc/nodesfile.json [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        "$type": "OpcPlc.ConfigNode, opcplc",
+        "NodeId": "1023",
```
Comment (hansgschossmann) [inline opcplc/nodesfile.json]: pls remove $type keys in each list element

File: opcplc/ConfigNode.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        CurrentWrite = 2,
+        CurrentReadOrWrite = 3,
+        HistoryRead = 4,
+        HistoryWrite = 8,
+        HistoryReadOrWrite = 12,
+        SemanticChange = 16,
+        StatusWrite = 32,
+        TimestampWrite = 64,
+    }
+}
```
Comment (hansgschossmann) [inline opcplc/ConfigNode.cs]: no longer needed if you do the other change to remove the switch

File: opcplc/ConfigNode.cs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace OpcPlc
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class ConfigFolder
+    {
+        public ushort NamespaceId { get; set; }
+
```
Comment (hansgschossmann) [inline opcplc/ConfigNode.cs]: not required

File: opcplc/ConfigNode.cs [hunk 16]
```diff
@@ -0,0 +1,52 @@
+﻿using Opc.Ua;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OpcPlc
+{
+    /// <summary>
+    /// 
+    /// </summary>
```
Comment (hansgschossmann) [inline opcplc/ConfigNode.cs]: comment missing

File: opcplc/PlcNodeManagerFromFile.cs [hunk 17]
```diff
@@ -0,0 +1,544 @@
...
+            }
+
+            return folder;
+        }
+
+        /// <summary>
+        /// Creates address space from JSON config file.
+        /// </summary>
+        /// <param name="externalReferences"></param>
+        public override void CreateAddressSpace(IDictionary<NodeId, IList<IReference>> externalReferences)
```
Comment (hansgschossmann) [inline opcplc/PlcNodeManagerFromFile.cs]: pls only use summary in method headers to be consistent with the other code

File: opcplc/PlcNodeManagerFromFile.cs [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+            //    },
+            //    TypeNameHandling = TypeNameHandling.All
+            //});
+
+            //using (StreamWriter sw = new StreamWriter("nodesfile.json"))
+            //{
+            //    sw.Write(json);
+            //}
+        }
+    }
```
Comment (hansgschossmann) [inline opcplc/PlcNodeManagerFromFile.cs]: unused, pls remove

File: opcplc/nodesfile.json [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        "NodeId": "1029",
+        "Name": "Active power",
+        "DataType": "Float",
+        "ValueRank": -1,
+        "AccessLevel": "CurrentReadOrWrite",
+        "Description": "Active power"
+      }
+    ]
+  }
+}
```
Comment (hansgschossmann) [inline opcplc/nodesfile.json]: here is an example with all modifications:
{
  "Name": "MyTelemetry",
  "NodeList": [
      {
        "NodeId": "1023",
        "Name": "ActualSpeed",
        "DataType": "Float",
        "ValueRank": -1,
        "AccessLevel": "CurrentReadOrWrite",
        "Description": "Rotational speed"
      },
      {
        "NodeId": "1024",
        "Name": "aRMS",
        "DataType": "Float",
        "ValueRank": -1,
        "AccessLevel": "CurrentReadOrWrite",
        "Description": "Mean of the vibration acceleration"
      },
      {
        "NodeId": "1025",
        "Name": "DKW",
        "DataType": "Float",
        "ValueRank": -1,
        "AccessLevel": "CurrentReadOrWrite",
        "Description": "Diagnostic characteristic value"
      },
      {
        "NodeId": "1026",
        "Name": "Peak value",
        "DataType": "Float",
        "ValueRank": -1,
        "AccessLevel": "CurrentReadOrWrite",
        "Description": "Peak value of vibration acceleration"
      },
      {
        "NodeId": "1027",
        "Name": "Vibration Velocity",
        "DataType": "Float",
        "ValueRank": -1,
        "AccessLevel": "CurrentReadOrWrite",
        "Description": "Mean of the vibration velocity"
      },
      {
        "NodeId": "1029",
        "Name": "Active power",
        "DataType": "Float",
        "ValueRank": -1,
        "AccessLevel": "CurrentReadOrWrite",
        "Description": "Active power"
      }
    ]
}

File: README.md (hunk 3 above)
Comment (ddobric) [reply README.md]: Corrected

File: opcplc/nodesfile.json (hunk 11 above)
Comment (ddobric) [reply opcplc/nodesfile.json]: removed

File: opcplc/ConfigNode.cs (hunk 14 above)
Comment (ddobric) [reply opcplc/ConfigNode.cs]: removed

**Pull request merged**